src/toolchain.ml

```diff
@@ -0,0 +1,240 @@
+open Import
+
+let get_opamroot () =
+  let open Promise.Syntax in
+  let+ home =
+    if not (Ocaml_windows.use_ocaml_env ()) then
+      Promise.return @@ Node.Os.homedir ()
+    else
+      let+ output = Ocaml_windows.cygwin_home () in
+      match output with
+      | Ok cygwin_home -> cygwin_home
+      | Error e ->
+        show_message `Warn
+          "Could not determine Cygwin home directory, defaulting to Windows \
+           home directory: %s"
+          e;
+        Node.Os.homedir ()
+  in
+  Path.(of_string home / ".vscode-ocaml" / "opam")
+
+type t =
+  { project_sandbox : Sandbox.t option
+  ; opam : Opam.t * Opam.Switch.t
+  ; version : string
+  }
+
+module Tool = struct
+  type t =
+    { opam_package : string
+    ; cmd : Cmd.t
+    }
+
+  let opam_package t = t.opam_package
+
+  let cmd t = t.cmd
+
+  let make ?(args = []) ~opam_package
+      { opam = opam, switch; project_sandbox; _ } bin =
+    match project_sandbox with
+    | Some sandbox ->
+      let open Promise.Syntax in
+      let+ has_command = Sandbox.has_command sandbox bin in
+      if has_command then
+        { opam_package; cmd = Sandbox.get_command sandbox bin args }
+      else
+        { opam_package; cmd = Opam.exec opam switch ~args:(bin :: args) }
+    | _ ->
+      Promise.return
+        { opam_package; cmd = Opam.exec opam switch ~args:(bin :: args) }
+
+  let lsp ?args toolchain =
+    make ~opam_package:"ocaml-lsp-server" toolchain "ocamllsp" ?args
+
+  let dune ?args toolchain = make ~opam_package:"dune" toolchain "dune" ?args
+
+  let ocamlformat ?args toolchain =
+    make ~opam_package:"ocamlformat" toolchain "ocamlformat" ?args
+
+  let all toolchain =
+    Promise.all_list
+      [ lsp toolchain ~args:[ "--version" ]
+      ; ocamlformat toolchain ~args:[ "--version" ]
+      ]
+
+  let detect_missing toolchain =
+    let open Promise.Syntax in
+    let missing_command_opt t =
+      let cmd = t.cmd in
+      let+ output =
+        let open Promise.Result.Syntax in
+        let* command = Cmd.check cmd in
+        Cmd.output command
+      in
+      match output with
+      | Ok _ -> None
+      | Error _ -> Some t
+    in
+    let* all = all toolchain in
+    Promise.List.filter_map missing_command_opt all
+end
+
+let opam_init opam opamroot =
+  let open Promise.Syntax in
+  let* exists = Fs.exists (opamroot |> Path.to_string) in
+  if exists then
+    Promise.Result.return ()
+  else
+    let+ result = Opam.init opam |> Cmd.output in
+    Result.map ~f:(fun _ -> ()) result
```

Comment:
`opam init` takes a bit of time on the computers I tried it on. Maybe it should display progress too?